File: web/src/domain/proxy/PacienteRepositoryProxy.ts
import { IPacienteRepository } from "../repositories/PacienteRepository";
import { Paciente } from "../models/Paciente";

export class PacienteRepositoryProxy implements IPacienteRepository {
  constructor(
    private realRepository: IPacienteRepository,
    private userRole: string
  ) {}

  async getAll(opticaId: string, options?: any): Promise<Paciente[]> {
    return this.realRepository.getAll(opticaId, options);
  }

  async getById(id: string, opticaId: string): Promise<Paciente | null> {
    return this.realRepository.getById(id, opticaId);
  }

  async save(paciente: Paciente): Promise<void> {
    return this.realRepository.save(paciente);
  }

  async delete(id: string, opticaId: string): Promise<void> {
    if (this.userRole !== "admin") {
      throw new Error("Acceso denegado: Solo administradores pueden eliminar.");
    }
    return this.realRepository.delete(id, opticaId);
  }
}

File: app/src/main/java/com/example/optoapp/domain/proxy/PacienteRepositoryProxy.kt
package com.example.optoapp.domain.proxy

import com.example.optoapp.domain.model.PacienteModel
import com.example.optoapp.domain.repository.PacienteRepository
import kotlinx.coroutines.flow.Flow

/**
 * Proxy para controlar el acceso al repositorio de pacientes.
 * Verifica permisos (roles) antes de delegar la llamada al repositorio real.
 */
class PacienteRepositoryProxy(
    private val realRepository: PacienteRepository,
    private val userRole: String
) : PacienteRepository {

    override fun getPacientes(opticaId: String): Flow<List<PacienteModel>> {
        return realRepository.getPacientes(opticaId)
    }

    override fun searchPacientes(opticaId: String, query: String): Flow<List<PacienteModel>> {
        return realRepository.searchPacientes(opticaId, query)
    }

    override suspend fun getPacienteById(id: String): PacienteModel? {
        return realRepository.getPacienteById(id)
    }

    override suspend fun insertPaciente(paciente: PacienteModel) {
        if (userRole == "asesor_ventas") {
            // Un asesor podría tener permiso limitado, pero aquí ejemplificamos la restricción
        }
        realRepository.insertPaciente(paciente)
    }

    override suspend fun updatePaciente(paciente: PacienteModel) {
        realRepository.updatePaciente(paciente)
    }

    override suspend fun deletePaciente(paciente: PacienteModel) {
        if (userRole != "admin") {
            throw SecurityException("Solo los administradores pueden eliminar pacientes.")
        }
        realRepository.deletePaciente(paciente)
    }
}
